File: go/2026-06-12-concurrent-task-runner/solution_test.go
package solution

import (
	"fmt"
	"testing"
)

type addTask struct {
	id int
	a  int
	b  int
}

func (t addTask) ID() int {
	return t.id
}

func (t addTask) Execute() int {
	return t.a + t.b
}

func Test(t *testing.T) {
	type testCase struct {
		name        string
		tasks       []addTask
		workerCount int
		expected    map[int]int
	}

	runCases := []testCase{
		{
			name: "small set, single worker",
			tasks: []addTask{
				{1, 2, 3},
				{2, 10, -5},
			},
			workerCount: 1,
			expected: map[int]int{
				1: 5,
				2: 5,
			},
		},
		{
			name: "three tasks, many workers",
			tasks: []addTask{
				{10, 100, 200},
				{11, -3, -7},
				{12, 42, 0},
			},
			workerCount: 5,
			expected: map[int]int{
				10: 300,
				11: -10,
				12: 42,
			},
		},
	}

	submitCases := append(runCases, []testCase{
		{
			name: "no tasks",
			tasks:       []addTask{},
			workerCount: 4,
			expected:    map[int]int{},
		},
		{
			name: "many tasks, worker limit smaller than tasks",
			tasks: []addTask{
				{1, 1, 1},
				{2, 2, 2},
				{3, 3, 3},
				{4, 4, 4},
				{5, 5, 5},
				{6, -10, 7},
				{7, 1000, -1},
			},
			workerCount: 3,
			expected: map[int]int{
				1: 2,
				2: 4,
				3: 6,
				4: 8,
				5: 10,
				6: -3,
				7: 999,
			},
		},
		{
			name: "unordered IDs, more workers than tasks",
			tasks: []addTask{
				{100, 7, 8},
				{2, 50, 50},
				{42, -100, 1},
			},
			workerCount: 10,
			expected: map[int]int{
				100: 15,
				2:   100,
				42:  -99,
			},
		},
	}...)

	testCases := runCases
	if withSubmit {
		testCases = submitCases
	}

	skipped := len(submitCases) - len(testCases)

	passCount := 0
	failCount := 0

	for _, test := range testCases {
		result := RunTasks[addTask](test.tasks, test.workerCount)

		if !mapsEqual(result, test.expected) {
			failCount++
			fmtTasks := formatTasks(test.tasks)
			t.Errorf(`---------------------------------
Test: %s
Worker count: %d
Input tasks:
%s
Expected results: %v
Actual results:   %v
Fail
`, test.name, test.workerCount, fmtTasks, test.expected, result)
		} else {
			passCount++
			fmtTasks := formatTasks(test.tasks)
			fmt.Printf(`---------------------------------
Test: %s
Worker count: %d
Input tasks:
%s
Expected results: %v
Actual results:   %v
Pass
`, test.name, test.workerCount, fmtTasks, test.expected, result)
		}
	}

	fmt.Println("---------------------------------")
	if skipped > 0 {
		fmt.Printf("%d passed, %d failed, %d skipped\n", passCount, failCount, skipped)
	} else {
		fmt.Printf("%d passed, %d failed\n", passCount, failCount)
	}
}

func mapsEqual(a, b map[int]int) bool {
	if len(a) != len(b) {
		return false
	}
	for k, v := range a {
		if b[k] != v {
			return false
		}
	}
	return true
}

func formatTasks(tasks []addTask) string {
	if len(tasks) == 0 {
		return "  (no tasks)\n"
	}

	result := ""
	for _, task := range tasks {
		result += fmt.Sprintf("  * ID: %d, a: %d, b: %d\n", task.id, task.a, task.b)
	}
	return result
}

// withSubmit is set at compile time depending
// on which button is used to run the tests
var withSubmit = true

File: Python/2026-08-09-Build-Overlapping-Tokens-Chunks/solution_test.py
from solution import chunk_tokens

run_cases = [
    (["a", "b", "c", "d", "e"], 3, 1, [["a", "b", "c"], ["c", "d", "e"]]),
    (
        ["one", "two", "three", "four", "five"],
        4,
        1,
        [["one", "two", "three", "four"], ["four", "five"]],
    ),
]

submit_cases = run_cases + [
    ([], 3, 1, []),
    (["a", "b"], 0, 0, ValueError),
    (["a", "b", "c"], 3, [-1, 3, 4], ValueError),
    (
        ["t0", "t1", "t2", "t3", "t4", "t5", "t6", "t7"],
        4,
        2,
        [
            ["t0", "t1", "t2", "t3"],
            ["t2", "t3", "t4", "t5"],
            ["t4", "t5", "t6", "t7"],
        ],
    ),
]


def test(tokens, chunk_size, overlap, expected):
    print("---------------------------------")
    print(f"Tokens:     {tokens}")
    print(f"Chunk size: {chunk_size}")
    print(f"Overlap:    {overlap}")
    print("")

    overlaps = overlap if isinstance(overlap, list) else [overlap]

    if expected is ValueError:
        for current_overlap in overlaps:
            try:
                result = chunk_tokens(tokens, chunk_size, current_overlap)
                print(f"Overlap {current_overlap}: expected ValueError")
                print(f"Actual: {result}")
                return False
            except ValueError as error:
                print(f"Overlap {current_overlap}: ValueError: {error}")
        return True

    result = chunk_tokens(tokens, chunk_size, overlap)
    print(f"Expected: {expected}")
    print(f"Actual:   {result}")
    return result == expected


def main():
    passed = 0
    failed = 0
    skipped = len(submit_cases) - len(test_cases)

    for test_case in test_cases:
        try:
            correct = test(*test_case)
        except Exception as error:
            print(f"Error: {error}")
            correct = False

        if correct:
            passed += 1
            print("Pass")
        else:
            failed += 1
            print("Fail")

    if failed == 0:
        print("============= PASS ==============")
    else:
        print("============= FAIL ==============")

    if skipped > 0:
        print(f"{passed} passed, {failed} failed, {skipped} skipped")
    else:
        print(f"{passed} passed, {failed} failed")


if __name__ == "__main__":
    test_cases = run_cases + submit_cases
    main()
